### ISSUE TITLE
False positive on union types for python 3.13

### ISSUE THREAD

**Issue by Anaritus** [astral-sh/ty#3168]
### Summary

With this main.py
```py
import sys

a: str | int


def main():
  print(sys.version)
  a: dict[str, str | int] = {}


if __name__ == "__main__":
  main()
```

ty check main.py yields this result
```
error[unsupported-operator]: Unsupported `|` operation
 --> main.py:4:4
  |
2 | import sys
3 |
4 | a: str | int
  |    ---^^^---
  |    |     |
  |    |     Has type `<class 'int'>`
  |    Has type `<class 'str'>`
  |
info: All type expressions are evaluated at runtime by default on Python <3.14
info: Python 3.13 was assumed when inferring types
  --> pyproject.toml:30:18
   |
29 | [tool.ty.environment]
30 | python-version = "3.13"
   |                  ^^^^^^ Python version configuration
31 | extra-paths = [
32 |   "/Users/som38/arcadia/contrib/python/PyYAML/py3",
   |
info: rule `unsupported-operator` is enabled by default

error[unsupported-operator]: Unsupported `|` operation
 --> main.py:9:16
  |
7 | def main():
8 |   print(sys.version)
9 |   a: dict[str, str | int] = {}
  |                ---^^^---
  |                |     |
  |                |     Has type `<class 'int'>`
  |                Has type `<class 'str'>`
  |
info: All type expressions are evaluated at runtime by default on Python <3.14
info: Python 3.13 was assumed when inferring types
  --> pyproject.toml:30:18
   |
29 | [tool.ty.environment]
30 | python-version = "3.13"
   |                  ^^^^^^ Python version configuration
31 | extra-paths = [
32 |   "/Users/som38/arcadia/contrib/python/PyYAML/py3",
   |
info: rule `unsupported-operator` is enabled by default
```

In the editor I also see that its unsupported-operator. But when I run the code it works. For example with
uv run -p 3.13 main.py

Should I suppress this error? What would be the best move here?

**Comment by Anaritus** [comment]
I think that is a false positive since it works with python as old as 3.10

**Comment by Anaritus** [comment]
Also I get this error aswell

<img width="2232" height="130" alt="Image" src="https://github.com/user-attachments/assets/dee2eb3c-8346-4b80-8674-5476cd1c0aa8" />

The info is
<img width="1271" height="536" alt="Image" src="https://github.com/user-attachments/assets/9a8aa025-cacc-4630-94e9-7a2c0dde07d5" />

For some reason it does not work as it should I think.

**Comment by Anaritus** [comment]
<img width="2095" height="1056" alt="Image" src="https://github.com/user-attachments/assets/c681f2aa-158b-4139-91e3-46fc1cb490c9" />

Also pydantic just does not work((

Im sorry if I should've done a separate issue for each thing that does not work, but I just tried ty and all the code I am inspecting just lit on fire

**Comment by Anaritus** [comment]
I think this is caused by my company's python env, but pyright works just fine(

**Comment by Anaritus** [comment]
disabling work based python executable in suggestions does work, but have no clue why. Even fixes the unsupported-operator diagnostic. Feel free to close the issue but Im open to try and debug it

**Comment by sharkdp** [comment]
Hm, interesting. I can not [reproduce this](https://play.ty.dev/a306abe2-7e37-46b3-ace1-4fbe20e612d4).

> I think this is caused by my company's python env

What exactly does that mean? What does that environment entail?

> disabling work based python executable in suggestions does work

what do you mean by "in suggestions"?

**Comment by Anaritus** [comment]
The company I work at does its own python, I do not know how specifically it differs from official builds, but what I know is that 3.13.2 work build does differ from the standard one. For example, this is the definition of Self from my company's build:

<img width="887" height="478" alt="Image" src="https://github.com/user-attachments/assets/733ee90e-f56a-4bf9-bcc3-581fb0dcc66d" />

And this is the deifinition of Self in "normal" build, the one ty uses by default.

<img width="1053" height="395" alt="Image" src="https://github.com/user-attachments/assets/55121993-367f-47bb-bff9-70e50db92221" />

Feel free to ask more clarifying questions, as english is not my first language. The definition of _SpecialForm looks exactly the same

**Comment by carljm** [comment]
@Anaritus That just looks like the runtime standard library vs typeshed (the type stubs used by all type checkers); it doesn't look related.

Can you reproduce any of these problems (let's focus on a simpler one, like the `Self` `_SpecialForm` error, or the union error) via `ty check somefile.py` on the command line? If so, please paste the output from `ty check -vv somefile.py`.

**Comment by dhruvmanila** [comment]
Can you provide the ty version used to check this?

I don't see this diagnostic on the [Playground](https://play.ty.dev/39f897e6-48b5-404d-bb37-8606b62153f6), it only occurs on 3.9 or lower (as expected) in the playground.

**Comment by Anaritus** [comment]
Sure thing! Replaced every directory and file with REDACTED for nda purposes
```
2026-03-31 14:56:45.137414 DEBUG Version: 0.0.26 (940305127 2026-03-26)
2026-03-31 14:56:45.142521 DEBUG Architecture: aarch64, OS: macos, case-sensitive: case-sensitive
2026-03-31 14:56:45.142537 DEBUG Searching for a project in REDACTED
2026-03-31 14:56:45.14285 DEBUG Found project at REDACTED
2026-03-31 14:56:45.14286 DEBUG Searching for a user-level configuration at REDACTED
2026-03-31 14:56:45.143463 INFO Defaulting to python-platform `darwin`
2026-03-31 14:56:45.143476 DEBUG Discovering virtual environment in REDACTED
2026-03-31 14:56:45.143576 DEBUG Attempting to parse virtual environment metadata at REDACTED
2026-03-31 14:56:45.143754 DEBUG Attempting to parse virtual environment metadata at REDACTED
2026-03-31 14:56:45.143771 DEBUG Searching for site-packages directory in sys.prefix REDACTED
2026-03-31 14:56:45.143789 DEBUG Failed to discover site-packages in ty's environment: Failed to iterate over the contents of the `lib`/`lib64` directories of the Python installation REDACTED
2026-03-31 14:56:45.143798 DEBUG Searching for site-packages directory in sys.prefix REDACTED
2026-03-31 14:56:45.143946 DEBUG Resolved site-packages directories for this virtual environment are: REDACTED
2026-03-31 14:56:45.144057 DEBUG Searching for real stdlib directory in sys.prefix REDACTED
2026-03-31 14:56:45.14407 DEBUG Resolved real stdlib path for this virtual environment is: REDACTED
2026-03-31 14:56:45.144236 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144294 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144328 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144373 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144414 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144451 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144488 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.14452 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144555 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144599 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144637 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144685 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144753 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144795 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144831 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144872 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144909 DEBUG Adding extra search-path REDACTED
2026-03-31 14:56:45.144914 DEBUG Adding first-party search path REDACTED
2026-03-31 14:56:45.144919 DEBUG Using vendored stdlib
2026-03-31 14:56:45.145162 DEBUG Adding site-packages search path REDACTED
2026-03-31 14:56:45.145182 INFO Python version: Python 3.13, platform: darwin
2026-03-31 14:56:45.145194 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145226 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145232 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.14524 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145244 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145248 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145253 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145257 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145261 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145266 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145281 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145286 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.14529 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145296 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.1453 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145304 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145308 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145356 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-03-31 14:56:45.145735 DEBUG Adding new file root REDACTED of kind Project
2026-03-31 14:56:45.145775 DEBUG Setting included paths: 1
2026-03-31 14:56:45.14578 DEBUG Reloading files for project `conf_parse`
2026-03-31 14:56:45.145815 DEBUG Starting main loop
2026-03-31 14:56:45.145826 DEBUG Waiting for next main loop message.
2026-03-31 14:56:45.145869 DEBUG Checking all files in project 'conf_parse'
2026-03-31 14:56:45.148989 INFO Indexed 1 file(s) in 0.003s
2026-03-31 14:56:45.149239 DEBUG Checking file REDACTED
2026-03-31 14:56:45.149654 DEBUG Resolving dynamic module resolution paths
2026-03-31 14:56:45.169085 DEBUG Module `__builtins__` not found in search paths
2026-03-31 14:56:45.180178 DEBUG Module `typing.type_check_only` not found in search paths
2026-03-31 14:56:45.180191 DEBUG Module `typing.type_check_only` not found while looking in parent dirs (neither stub nor real module file)
2026-03-31 14:56:45.183036 DEBUG Module `collections.abc` not found in search paths
2026-03-31 14:56:45.183042 DEBUG Module `collections.abc` not found while looking in parent dirs (neither stub nor real module file)
2026-03-31 14:56:45.185577 DEBUG Module `collections.abc` not found while looking in parent dirs (neither stub nor real module file)
2026-03-31 14:56:45.185619 DEBUG Module `_typing` not found in search paths
2026-03-31 14:56:45.185623 DEBUG Module `_typing` not found while looking in parent dirs (neither stub nor real module file)
2026-03-31 14:56:45.199675 DEBUG Module `_typing` not found while looking in parent dirs (neither stub nor real module file)
2026-03-31 14:56:45.202095 DEBUG Module `collections.abc` not found while looking in parent dirs (neither stub nor real module file)
2026-03-31 14:56:45.213386 DEBUG Module `collections.abc` not found while looking in parent dirs (neither stub nor real module file)
2026-03-31 14:56:45.222858 INFO Error looking up `typing.Sequence` in typeshed: expected to find a class definition on Python 3.13, but found a symbol of type `_SpecialGenericAlias` instead. Falling back to `Unknown` for the symbol instead.
2026-03-31 14:56:45.228387 DEBUG Module `_sre` not found in search paths
2026-03-31 14:56:45.228396 DEBUG Module `_sre` not found while looking in parent dirs (neither stub nor real module file)
2026-03-31 14:56:45.23459 INFO Error looking up `typing.TypeVar` in typeshed: expected to find a class definition on Python 3.13, but found a symbol of type `Unknown` instead. Falling back to `Unknown` for the symbol instead.
2026-03-31 14:56:45.23504 DEBUG Checking all files took 0.086s
error[unsupported-operator]: Unsupported `|` operation
  --> src/models/mobile_button_entity.py:22:12
   |
21 | class SpecObject(BaseModel):
22 |   version: str | None = None
   |            ---^^^----
   |            |     |
   |            |     Has type `None`
   |            Has type `<class 'str'>`
23 |   api_key: int
24 |   status: SpecStatus
   |
info: All type expressions are evaluated at runtime by default on Python <3.14
info: Python 3.13 was assumed when inferring types
  --> ty.toml:22:18
   |
20 |   REDACTED
21 | ]
22 | python-version = "3.13"
   |                  ^^^^^^ Python version configuration
   |
info: rule `unsupported-operator` is enabled by default

This is output up to the first diagnostic, inclusively, which happens to be unsupported operator one.
```
Also
λ ty -V
ty 0.0.26 (940305127 2026-03-26)

**Comment by carljm** [comment]
Ok, maybe your previous comment was more relevant than I thought. It looks likely to me that one of the library search paths being added (probably via a `PYTHONPATH` environment variable) is a copy of either typeshed or the Python standard library, which is taking priority over the vendored typeshed, and is either for the wrong Python version, or just contains a bunch of stuff that breaks our understanding of fairly common types. Maybe try again with `PYTHONPATH=`? Or find the `PYTHONPATH` entry that looks like it might be stdlib/typeshed and remove that one?

**Comment by Anaritus** [comment]
Yeah when I remove the extra-path that contains stdlib it works, and Im using it currently. This is with PYTHONPATH= (without removing the stdlib extra-path):
```
λ export PYTHONPATH= && ty check -vv src/models/mobile_button_entity.py
2026-04-01 20:28:25.161416 DEBUG Version: 0.0.26 (940305127 2026-03-26)
2026-04-01 20:28:25.162277 DEBUG Architecture: aarch64, OS: macos, case-sensitive: case-sensitive
2026-04-01 20:28:25.162289 DEBUG Searching for a project in REDACTED
2026-04-01 20:28:25.162459 DEBUG Found project at REDACTED
2026-04-01 20:28:25.162468 DEBUG Searching for a user-level configuration at `REDACTED
2026-04-01 20:28:25.162745 INFO Defaulting to python-platform `darwin`
2026-04-01 20:28:25.162752 DEBUG Discovering virtual environment in `REDACTED
2026-04-01 20:28:25.162802 DEBUG Attempting to parse virtual environment metadata at REDACTED
2026-04-01 20:28:25.162915 DEBUG Attempting to parse virtual environment metadata at REDACTED
2026-04-01 20:28:25.162921 DEBUG Searching for site-packages directory in sys.prefix REDACTED
2026-04-01 20:28:25.162928 DEBUG Failed to discover site-packages in ty's environment: Failed to iterate over the contents of the `lib`/`lib64` directories of the Python installation `REDACTED
2026-04-01 20:28:25.162932 DEBUG Searching for site-packages directory in sys.prefix REDACTED
2026-04-01 20:28:25.163012 DEBUG Resolved site-packages directories for this virtual environment are: ["REDACTED
2026-04-01 20:28:25.163062 DEBUG Searching for real stdlib directory in sys.prefix REDACTED
2026-04-01 20:28:25.163068 DEBUG Resolved real stdlib path for this virtual environment is: REDACTED
2026-04-01 20:28:25.16313 DEBUG Adding `REDACTED from the `PYTHONPATH` environment variable to `extra_paths`
2026-04-01 20:28:25.163159 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163188 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163209 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163229 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163256 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163278 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163298 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163316 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163338 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.16336 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163381 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.1634 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163424 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163446 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163465 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163485 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163505 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163525 DEBUG Adding extra search-path `REDACTED
2026-04-01 20:28:25.163528 DEBUG Adding first-party search path `REDACTED
2026-04-01 20:28:25.163531 DEBUG Using vendored stdlib
2026-04-01 20:28:25.163654 DEBUG Adding site-packages search path `REDACTED
2026-04-01 20:28:25.163668 INFO Python version: Python 3.13, platform: darwin
2026-04-01 20:28:25.163675 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.16369 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163696 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.1637 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163703 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163707 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163711 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163715 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163718 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163723 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163726 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163729 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163741 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163744 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163748 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163751 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163754 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163758 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163764 DEBUG Adding new file root REDACTED of kind LibrarySearchPath
2026-04-01 20:28:25.163985 DEBUG Setting included paths: 1
2026-04-01 20:28:25.163991 DEBUG Reloading files for project `conf_parse`
2026-04-01 20:28:25.164007 DEBUG Starting main loop
2026-04-01 20:28:25.164012 DEBUG Waiting for next main loop message.
2026-04-01 20:28:25.164033 DEBUG Checking all files in project 'conf_parse'
2026-04-01 20:28:25.166814 INFO Indexed 1 file(s) in 0.003s
2026-04-01 20:28:25.167024 DEBUG Checking file REDACTED
2026-04-01 20:28:25.167356 DEBUG Resolving dynamic module resolution paths
2026-04-01 20:28:25.183156 DEBUG Module `__builtins__` not found in search paths
2026-04-01 20:28:25.193019 DEBUG Module `typing.type_check_only` not found in search paths
2026-04-01 20:28:25.193037 DEBUG Module `typing.type_check_only` not found while looking in parent dirs (neither stub nor real module file)
2026-04-01 20:28:25.195857 DEBUG Module `collections.abc` not found in search paths
2026-04-01 20:28:25.195866 DEBUG Module `collections.abc` not found while looking in parent dirs (neither stub nor real module file)
2026-04-01 20:28:25.198549 DEBUG Module `collections.abc` not found while looking in parent dirs (neither stub nor real module file)
2026-04-01 20:28:25.198596 DEBUG Module `_typing` not found in search paths
2026-04-01 20:28:25.198599 DEBUG Module `_typing` not found while looking in parent dirs (neither stub nor real module file)
2026-04-01 20:28:25.211985 DEBUG Module `_typing` not found while looking in parent dirs (neither stub nor real module file)
2026-04-01 20:28:25.214562 DEBUG Module `collections.abc` not found while looking in parent dirs (neither stub nor real module file)
2026-04-01 20:28:25.226774 DEBUG Module `collections.abc` not found while looking in parent dirs (neither stub nor real module file)
2026-04-01 20:28:25.236683 INFO Error looking up `typing.Sequence` in typeshed: expected to find a class definition on Python 3.13, but found a symbol of type `_SpecialGenericAlias` instead. Falling back to `Unknown` for the symbol instead.
2026-04-01 20:28:25.24241 DEBUG Module `_sre` not found in search paths
2026-04-01 20:28:25.242431 DEBUG Module `_sre` not found while looking in parent dirs (neither stub nor real module file)
2026-04-01 20:28:25.249684 INFO Error looking up `typing.TypeVar` in typeshed: expected to find a class definition on Python 3.13, but found a symbol of type `Unknown` instead. Falling back to `Unknown` for the symbol instead.
2026-04-01 20:28:25.250172 DEBUG Checking all files took 0.083s
error[unsupported-operator]: Unsupported `|` operation
  --> src/models/mobile_button_entity.py:22:12
   |
21 | class SpecObject(BaseModel):
22 |   version: str | None = None
   |            ---^^^----
   |            |     |
   |            |     Has type `None`
   |            Has type `<class 'str'>`
23 |   api_key: int
24 |   status: SpecStatus
   |
info: All type expressions are evaluated at runtime by default on Python <3.14
info: Python 3.13 was assumed when inferring types
  --> ty.toml:22:18
   |
20 |   "REDACTED
21 | ]
22 | python-version = "3.13"
   |                  ^^^^^^ Python version configuration
   |
info: rule `unsupported-operator` is enabled by default
```

I guess what happens is it reads the stdlib extra-path gets read and overrides the env variable (or makes it irrelevant)

**Comment by carljm** [comment]
I don't think it is expected for ty (or any type checker) to work correctly if the actual stdlib is included as an extra-path. Closing, since it sounds like that is the problem. Thanks for helping track it down!
